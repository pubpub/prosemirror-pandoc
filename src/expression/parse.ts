import { Expr } from "./types";

// A simple recursive-descent parser to turn expressions like `(Space | Str)+` into syntax trees
export const parseExpr = (str: string): Expr => {
    str = str.trim();
    // Remove spaces around choice separators
    str = str.replace(/\s*\|\s*/g, "|");
    // Remove extraneous spaces
    str = str.replace(/\s+/g, " ");

    // Keep track of the separator type we have at this level. It should either be choice separators
    // ("|") or sequential separators (" ").
    let separator;
    // Find separators
    const separators = [];
    // Keep track of open and close parens
    let parenCount = 0;
    // Keep track of open and close curlies
    let curlyCount = 0;
    for (let ptr = 0; ptr < str.length; ++ptr) {
        const char = str.charAt(ptr);
        if (char === "(") {
            ++parenCount;
        } else if (char === ")") {
            --parenCount;
        } else if (char === "{") {
            ++curlyCount;
        } else if (char === "}") {
            --curlyCount;
        } else if (
            parenCount === 0 &&
            curlyCount === 0 &&
            (char === "|" || char === " ")
        ) {
            if (separator && separator !== char) {
                throw new Error(
                    "Please surround mixed separators with parentheses!" +
                        ` e.g. prefer '(Foo | Bar) Baz' over 'Foo | Bar Baz'. (at ${ptr}, parsing '${str}')`
                );
            } else {
                separator = char;
                separators.push(ptr);
            }
        }
    }
    if (separators.length > 0) {
        const separated: string[] = [];
        let substring = "";
        for (let ptr = 0; ptr < str.length; ++ptr) {
            if (separators.includes(ptr)) {
                separated.push(substring);
                substring = "";
            } else {
                substring += str.charAt(ptr);
            }
        }
        if (substring.length > 0) {
            separated.push(substring);
        }
        return {
            type: separator === " " ? "sequence" : "choice",
            children: separated.map(parseExpr),
        };
    } else if (str.endsWith("}")) {
        let ptr = str.length - 1;
        while (str.charAt(ptr) !== "{") {
            ptr--;
        }
        const rangeStrs = str.slice(ptr + 1, str.length - 1).split(",");
        const hasTwo = rangeStrs.length === 2;
        const range = rangeStrs.map((str) => parseInt(str.trim()));
        const [lowerBound, upperBound] = range;
        return {
            type: "range",
            lowerBound,
            upperBound: hasTwo
                ? isNaN(upperBound)
                    ? null
                    : upperBound
                : lowerBound,
            child: parseExpr(str.slice(0, ptr)),
        };
    } else if (str.endsWith("+")) {
        return {
            type: "oneOrMore",
            child: parseExpr(str.slice(0, str.length - 1)),
        };
    } else if (str.endsWith("*")) {
        return {
            type: "zeroOrMore",
            child: parseExpr(str.slice(0, str.length - 1)),
        };
    } else if (str.startsWith("(") && str.endsWith(")")) {
        return parseExpr(str.slice(1, str.length - 1));
    }
    return { type: "identifier", identifier: str };
};
