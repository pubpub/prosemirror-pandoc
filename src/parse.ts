import {
    Alignment,
    Attr,
    Block,
    BlockQuote,
    BulletList,
    Caption,
    Cell,
    CitationMode,
    Cite,
    Code,
    CodeBlock,
    ColSpec,
    DefinitionList,
    Div,
    Doc,
    Format,
    Header,
    Image,
    Inline,
    LineBlock,
    Link,
    ListAttributes,
    ListNumberDelim,
    ListNumberStyle,
    Math,
    MathType,
    MetaBlocks,
    MetaBool,
    MetaInlines,
    MetaList,
    MetaMap,
    MetaString,
    MetaValue,
    Note,
    OrderedList,
    PandocJson,
    Para,
    Plain,
    Quoted,
    QuoteType,
    RawBlock,
    RawInline,
    Row,
    SimpleInline,
    Span,
    Str,
    Table,
    TableBody,
    TableFoot,
    TableHead,
    Target,
} from "./types";

const unwrapEnum = <T>(instance: { t: T }): T => {
    return instance.t;
};

const unwrapAttr = (attr: [string, string[], [string, string][]]): Attr => {
    const [identifier, classes, propertiesList] = attr;
    const properties = {};
    propertiesList.forEach(([key, value]) => {
        properties[key] = value;
    });
    return {
        identifier,
        classes,
        properties,
    };
};

const unwrapTarget = (target: [string, string]): Target => {
    const [url, title] = target;
    return {
        url,
        title,
    };
};

const unwrapFormat = (format: any): Format => {
    // TODO(ian): Figure out what to do here
    return format;
};

const unwrapListAttributes = (
    listAttributes: [number, any, any]
): ListAttributes => {
    const [startNumber = 1, listNumberStyle, listNumberDelim] = listAttributes;
    return {
        startNumber,
        listNumberStyle: unwrapEnum<ListNumberStyle>(listNumberStyle),
        listNumberDelim: unwrapEnum<ListNumberDelim>(listNumberDelim),
    };
};

const parseAtom = (n: { t }) => {
    return { type: n.t };
};

const parseStr = (n: { c: string }): Str => {
    const string = n.c;
    return {
        type: "Str",
        content: string,
    };
};

const parseSimpleInline = (
    n: {
        c: any[];
    },
    nodeType: SimpleInline["type"]
): SimpleInline => {
    const inline = n.c;
    return {
        type: nodeType,
        content: inline.map(parseInline),
    };
};

const parseQuoted = (n: { c: [any, any[]] }): Quoted => {
    const [quoteType, inline] = n.c;
    return {
        type: "Quoted",
        quoteType: unwrapEnum<QuoteType>(quoteType),
        content: inline.map(parseInline),
    };
};

const parseCite = (n: { c: [any[], any[]] }): Cite => {
    const [citations, inline] = n.c;
    return {
        type: "Cite",
        citations: citations.map((citation) => {
            const {
                citationHash,
                citationId,
                citationMode,
                citationNoteNum,
                citationPrefix,
                citationSuffix,
            } = citation;
            return {
                citationHash,
                citationId,
                citationMode: unwrapEnum<CitationMode>(citationMode),
                citationNoteNum,
                citationPrefix: citationPrefix.map(parseInline),
                citationSuffix: citationSuffix.map(parseInline),
            };
        }),
        content: inline.map(parseInline),
    };
};

const parseCode = (n: { c: [any, string] }): Code => {
    const [attr, code] = n.c;
    return {
        type: "Code",
        attr: unwrapAttr(attr),
        content: code,
    };
};

const parseMath = (n: { c: [any, string] }): Math => {
    const [mathType, content] = n.c;
    return {
        type: "Math",
        mathType: unwrapEnum<MathType>(mathType),
        content: content,
    };
};

const parseRawInline = (n: { c: [any, string] }): RawInline => {
    const [format, content] = n.c;
    return {
        type: "RawInline",
        format: unwrapFormat(format),
        content,
    };
};

const parseImage = (n: { c: [any, any[], any] }): Image => {
    const [attr, inline, target] = n.c;
    return {
        type: "Image",
        attr: unwrapAttr(attr),
        content: inline.map(parseInline),
        target: unwrapTarget(target),
    };
};

const parseLink = (n: { c: [any, any[], any] }): Link => {
    const [attr, inline, target] = n.c;
    return {
        type: "Link",
        attr: unwrapAttr(attr),
        content: inline.map(parseInline),
        target: unwrapTarget(target),
    };
};

const parseNote = (n: { c: any[] }): Note => {
    const blocks = n.c;
    return {
        type: "Note",
        content: blocks.map(parseBlock),
    };
};

const parseSpan = (n: { c: [any, any[]] }): Span => {
    const [attr, inline] = n.c;
    return {
        type: "Span",
        attr: unwrapAttr(attr),
        content: inline.map(parseInline),
    };
};

export const parseInline = (n: { t: Inline["type"]; c: any }): Inline => {
    switch (n.t) {
        case "Str":
            return parseStr(n);
        case "Emph":
        case "Strong":
        case "Underline":
        case "Strikeout":
        case "Superscript":
        case "Subscript":
        case "SmallCaps":
            return parseSimpleInline(n, n.t);
        case "Quoted":
            return parseQuoted(n);
        case "Cite":
            return parseCite(n);
        case "Code":
            return parseCode(n);
        case "Space":
        case "SoftBreak":
        case "LineBreak":
            return parseAtom(n);
        case "Math":
            return parseMath(n);
        case "RawInline":
            return parseRawInline(n);
        case "Link":
            return parseLink(n);
        case "Image":
            return parseImage(n);
        case "Note":
            return parseNote(n);
        case "Span":
            return parseSpan(n);
    }
};

const parsePlain = (n: { c: any[] }): Plain => {
    const inline = n.c;
    return {
        type: "Plain",
        content: inline.map(parseInline),
    };
};

const parsePara = (n: { c: any[] }): Para => {
    const inline = n.c;
    return {
        type: "Para",
        content: inline.map(parseInline),
    };
};

const parseLineBlock = (n: { c: any[][] }): LineBlock => {
    const lines = n.c;
    return {
        type: "LineBlock",
        content: lines.map((line) => line.map((inline) => parseInline(inline))),
    };
};

const parseCodeBlock = (n: { c: [any, string] }): CodeBlock => {
    const [attr, content] = n.c;
    return {
        type: "CodeBlock",
        attr: unwrapAttr(attr),
        content,
    };
};

const parseRawBlock = (n: { c: [any, string] }): RawBlock => {
    const [format, content] = n.c;
    return {
        type: "RawBlock",
        format: unwrapFormat(format),
        content,
    };
};

const parseBlockQuote = (n: { c: any[] }): BlockQuote => {
    const blocks = n.c;
    return {
        type: "BlockQuote",
        content: blocks.map(parseBlock),
    };
};

const parseOrderedList = (n: { c: [any, any[][]] }): OrderedList => {
    const [listAttributes, items] = n.c;
    return {
        type: "OrderedList",
        listAttributes: unwrapListAttributes(listAttributes),
        content: items.map((item) => item.map(parseBlock)),
    };
};

const parseBulletList = (n: { c: any[][] }): BulletList => {
    const items = n.c;
    return {
        type: "BulletList",
        content: items.map((item) => item.map(parseBlock)),
    };
};

const parseDefinitionList = (n: { c: [any[], any[][]][] }): DefinitionList => {
    const items = n.c;
    const entries = items.map((item) => {
        const [term, definitions] = item;
        return {
            term: term.map(parseInline),
            definitions: definitions.map((definition) =>
                definition.map(parseBlock)
            ),
        };
    });
    return {
        type: "DefinitionList",
        entries,
    };
};

const parseHeader = (n: { c: [number, any, any[]] }): Header => {
    const [level, attr, inline] = n.c;
    return {
        type: "Header",
        level,
        attr: unwrapAttr(attr),
        content: inline.map(parseInline),
    };
};

const parseDiv = (n: { c: [any, any[]] }): Div => {
    const [attr, blocks] = n.c;
    return {
        type: "Div",
        attr: unwrapAttr(attr),
        content: blocks.map(parseBlock),
    };
};

const parseCell = (n: [any, any, any, any, any[]]): Cell => {
    const [attr, alignment, rowSpan, colSpan, blocks] = n;
    return {
        type: "Cell",
        attr: unwrapAttr(attr),
        alignment: unwrapEnum<Alignment>(alignment),
        rowSpan,
        colSpan,
        content: blocks.map(parseBlock),
    };
};

const parseRow = (n: [any, any[]]): Row => {
    const [attr, cells] = n;
    return {
        type: "Row",
        attr: unwrapAttr(attr),
        cells: cells.map(parseCell),
    };
};

const parseTableHead = (n: [any, any[]]): TableHead => {
    const [attr, rows] = n;
    return {
        type: "TableHead",
        attr: unwrapAttr(attr),
        rows: rows.map(parseRow),
    };
};

const parseTableFoot = (n: [any, any[]]): TableFoot => {
    const [attr, rows] = n;
    return {
        type: "TableFoot",
        attr: unwrapAttr(attr),
        rows: rows.map(parseRow),
    };
};

const parseTableBody = (n: [any, any, any[], any[]]): TableBody => {
    const [attr, rowHeadColumns, head, body] = n;
    return {
        type: "TableBody",
        rowHeadColumns,
        attr: unwrapAttr(attr),
        headRows: head.map(parseRow),
        bodyRows: body.map(parseRow),
    };
};

const parseColSpec = (n: [any, any]): ColSpec => {
    const [alignment, colWidth] = n;
    const base = {
        type: "ColSpec" as const,
        alignment: unwrapEnum<Alignment>(alignment),
    };
    if (colWidth.t === "ColWidthDefault") {
        return { ...base, defaultWidth: true };
    }
    return { ...base, width: colWidth.c };
};

const parseCaption = (n: [null | any[], any[]]): Caption => {
    const [shortCaption, content] = n;
    const baseCaption: Caption = {
        type: "Caption",
        content: content.map(parseBlock),
    };
    if (shortCaption) {
        return {
            ...baseCaption,
            shortCaption: shortCaption.map(parseInline),
        };
    }
    return baseCaption;
};

const parseTable = (n: { c: [any, any, any[], any, any[], any] }): Table => {
    const [attr, caption, colSpecs, head, bodies, foot] = n.c;
    return {
        type: "Table",
        attr: unwrapAttr(attr),
        caption: parseCaption(caption),
        colSpecs: colSpecs.map(parseColSpec),
        head: parseTableHead(head),
        bodies: bodies.map(parseTableBody),
        foot: parseTableFoot(foot),
    };
};

export const parseBlock = (n: { t: Block["type"]; c: any }): Block => {
    switch (n.t) {
        case "Plain":
            return parsePlain(n);
        case "Para":
            return parsePara(n);
        case "LineBlock":
            return parseLineBlock(n);
        case "CodeBlock":
            return parseCodeBlock(n);
        case "RawBlock":
            return parseRawBlock(n);
        case "BlockQuote":
            return parseBlockQuote(n);
        case "OrderedList":
            return parseOrderedList(n);
        case "BulletList":
            return parseBulletList(n);
        case "DefinitionList":
            return parseDefinitionList(n);
        case "Header":
            return parseHeader(n);
        case "HorizontalRule":
        case "Null":
            return parseAtom(n);
        case "Div":
            return parseDiv(n);
        case "Table":
            return parseTable(n);
    }
};

const parseMetaMap = (n: { c: { [key: string]: any } }): MetaMap => {
    const values = {};
    Object.entries(n.c).forEach(([key, value]) => {
        values[key] = parseMetaValue(value);
    });
    return { type: "MetaMap", values };
};

const parseMetaList = (n: { c: any[] }): MetaList => {
    return {
        type: "MetaList",
        content: n.c.map(parseMetaValue),
    };
};

const parseMetaBool = (n: { c: boolean }): MetaBool => {
    return {
        type: "MetaBool",
        content: n.c,
    };
};

const parseMetaString = (n: { c: string }): MetaString => {
    return {
        type: "MetaString",
        content: n.c,
    };
};

const parseMetaInlines = (n: { c: any[] }): MetaInlines => {
    return {
        type: "MetaInlines",
        content: n.c.map(parseInline),
    };
};

const parseMetaBlocks = (n: { c: any[] }): MetaBlocks => {
    return {
        type: "MetaBlocks",
        content: n.c.map(parseBlock),
    };
};

const parseMetaValue = (n: { t: string; c: any }): MetaValue => {
    switch (n.t) {
        case "MetaMap":
            return parseMetaMap(n);
        case "MetaList":
            return parseMetaList(n);
        case "MetaBool":
            return parseMetaBool(n);
        case "MetaString":
            return parseMetaString(n);
        case "MetaInlines":
            return parseMetaInlines(n);
        case "MetaBlocks":
            return parseMetaBlocks(n);
    }
};

const parseMeta = (meta: PandocJson["meta"]) => {
    const parsedMeta: { [key: string]: MetaValue } = {};
    Object.entries(meta).forEach(([key, value]) => {
        parsedMeta[key] = parseMetaValue(value);
    });
    return parsedMeta;
};

export const parsePandocJson = (json: PandocJson): Doc => {
    const { meta, blocks } = json;
    return {
        type: "Doc",
        blocks: blocks.map(parseBlock),
        meta: parseMeta(meta),
    };
};
