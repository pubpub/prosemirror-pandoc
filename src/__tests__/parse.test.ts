/* global describe, it, expect */
import { parseInline, parseBlock } from "../parse";

describe("parseInline", () => {
    it("parses a Str", () => {
        expect(parseInline({ t: "Str", c: "Hello!" })).toEqual({
            type: "Str",
            content: "Hello!",
        });
    });

    it("parses a simple inline elements", () => {
        [
            "Emph",
            "Strong",
            "Strikeout",
            "Superscript",
            "Subscript",
            "SmallCaps",
        ].forEach(type => {
            expect(
                // @ts-ignore
                parseInline({ t: type, c: [{ t: "Str", c: "Testing!" }] })
            ).toEqual({
                type,
                content: [{ type: "Str", content: "Testing!" }],
            });
        });
    });

    it("parses an atom element", () => {
        // @ts-ignore
        expect(parseInline({ t: "LineBreak" })).toEqual({ type: "LineBreak" });
    });

    it("parses a Link", () => {
        expect(
            parseInline({
                t: "Link",
                c: [
                    [
                        "this-is-an-identifier",
                        ["these", "are", "classes"],
                        [["key1", "val1"], ["key2", "val2"]],
                    ],
                    [
                        {
                            t: "Strong",
                            c: [
                                { t: "Str", c: "It's" },
                                { t: "Space" },
                                { t: "Str", c: "a" },
                                { t: "Space" },
                                { t: "Str", c: "link!" },
                            ],
                        },
                    ],
                    ["https://url.com", "This is a title??"],
                ],
            })
        ).toMatchSnapshot();
    });
});

describe("parseBlock", () => {
    it("parses a Para with some stuff in it", () => {
        expect(
            parseBlock({
                t: "Para",
                c: [
                    { t: "Strong", c: [{ t: "Str", c: "Hello," }] },
                    { t: "Space" },
                    { t: "Str", c: "world!" },
                ],
            })
        ).toEqual({
            type: "Para",
            content: [
                {
                    type: "Strong",
                    content: [{ type: "Str", content: "Hello," }],
                },
                { type: "Space" },
                { type: "Str", content: "world!" },
            ],
        });
    });

    it("parses an OrderedList", () => {
        expect(
            parseBlock({
                t: "OrderedList",
                c: [
                    [3, { t: "Decimal" }, { t: "Period" }],
                    [
                        [
                            {
                                t: "Para",
                                c: [{ t: "Str", c: "One!" }],
                            },
                        ],
                        [
                            {
                                t: "Para",
                                c: [{ t: "Str", c: "Two!" }],
                            },
                        ],
                        [
                            {
                                t: "Para",
                                c: [{ t: "Str", c: "Three!" }],
                            },
                        ],
                    ],
                ],
            })
        ).toMatchSnapshot();
    });
});
