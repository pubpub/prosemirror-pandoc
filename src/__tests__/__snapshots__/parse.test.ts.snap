// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseBlock parses an OrderedList 1`] = `
Object {
  "content": Array [
    Array [
      Object {
        "content": Array [
          Object {
            "content": "One!",
            "type": "Str",
          },
        ],
        "type": "Para",
      },
    ],
    Array [
      Object {
        "content": Array [
          Object {
            "content": "Two!",
            "type": "Str",
          },
        ],
        "type": "Para",
      },
    ],
    Array [
      Object {
        "content": Array [
          Object {
            "content": "Three!",
            "type": "Str",
          },
        ],
        "type": "Para",
      },
    ],
  ],
  "listAttributes": Object {
    "listNumberDelim": "Period",
    "listNumberStyle": "Decimal",
    "startNumber": 3,
  },
  "type": "OrderedList",
}
`;

exports[`parseInline parses a Link 1`] = `
Object {
  "attr": Object {
    "classes": Array [
      "these",
      "are",
      "classes",
    ],
    "identifier": "this-is-an-identifier",
    "properties": Object {
      "key1": "val1",
      "key2": "val2",
    },
  },
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": "It's",
          "type": "Str",
        },
        Object {
          "type": "Space",
        },
        Object {
          "content": "a",
          "type": "Str",
        },
        Object {
          "type": "Space",
        },
        Object {
          "content": "link!",
          "type": "Str",
        },
      ],
      "type": "Strong",
    },
  ],
  "target": Object {
    "title": "This is a title??",
    "url": "https://url.com",
  },
  "type": "Link",
}
`;
